### PR TITLE
Closes #2406 Delete "Remove Query Strings" feature

### LINKED ISSUE #2406: Get rid of the Remove Query Strings option
## **What are we proposing to do?**

With this update we are hoping to remove the option from WP Rocket UI: Remove query strings from static resources (File Optimization tab)

How our current option looks like: 

## **Why are we doing this?**

Ressource: <https://sirv.com/help/resources/remove-query-strings-from-static-resources/>

Here is a GTmetrix test comparison: 

- With query strings: <https://gtmetrix.com/reports/mega.wp-rocket.me/lXacNRNS>
- Without query strings: <https://gtmetrix.com/reports/mega.wp-rocket.me/MCZq51yj>

The scores are exactly the same. GTmetrix no longer recommends this nor takes into account query strings for the grades. 

## **Does query strings affect performance?**

Here are the headers for a second visit to the same page: 

Modern browsers and proxies cache based on headers and not based on the presence of query strings. 

## **User expectations and experience. How will we do this feature?**
From version 3.6, all users will not have this option in WP Rocket UI. 

Both existing and new users will have the same experience. Query strings will come back. But as we know from our research, this isn't going to affect performance. 

Pros:

- Simpler approach
- Easier to implement and maintain

Cons: 

- Higher risk since we don't know what edge cases we are missing.

### Sub-Cards
- [x] [3rd party part](https://github.com/wp-media/wp-rocket/issues/2436)
- [x] [Admin misc part](https://github.com/wp-media/wp-rocket/issues/2433)
- [x] [Facebook & Google Tracking part](https://github.com/wp-media/wp-rocket/issues/2438)
- [x] [Front part](https://github.com/wp-media/wp-rocket/issues/2434)
- [x] [Simple Custom CSS part](https://github.com/wp-media/wp-rocket/issues/2437)
- [x] [Common part](https://github.com/wp-media/wp-rocket/issues/2435)
- [x] [Settings part](https://github.com/wp-media/wp-rocket/issues/2432)
- [x] [Option management part](https://github.com/wp-media/wp-rocket/issues/2431)

## **Risks**

Since we are removing a feature, this should be low risk. The major concern is educating customers. 

## **Testing**

The following is written assuming the second approach is used where we remove query strings for existing and new users. 

**Acceptance Criteria:**

- Update WP Rocket to 3.6. (Existing install had the option "Remove Query Strings" enabled)
- The UI for "Remove Query Strings" is removed. 
- Assets should now have the query strings added to it. 
- When Minify CSS / Minify JS is enabled, these will also have the query strings. 

**What if:** 

We need to think of any possible edge cases. Can't think of anything at the time of writing this. 

## **Documentation and Translations**

- <https://docs.wp-rocket.me/article/56-remove-query-string-from-static-resources>
- <https://wp-rocket.me/features/> (Remove Query Strings is listed as a feature)
- Helper plugin: <https://github.com/wp-media/wp-rocket-helpers/tree/master/static-files/wp-rocket-static-exclude-query-string>

### PR DESCRIPTION
**Pull request by Tabrisrp**
Closes #2406 

Final PR with all the sub-tasks in, to merge in the 3.6 branch

### CODE REVIEW

File: inc/Engine/HealthCheck/CacheDirSizeCheck.php [hunk 1]
```diff
@@ -89,8 +88,7 @@ public function cache_dir_size_check() {
 		$current_blog_id = get_current_blog_id();
 
 		$checks = [
-			'min'     => WP_ROCKET_MINIFY_CACHE_PATH . $current_blog_id,
-			'busting' => WP_ROCKET_CACHE_BUSTING_PATH . $current_blog_id,
+			'min' => rocket_get_constant( 'WP_ROCKET_MINIFY_CACHE_PATH' ) . $current_blog_id,
```
Comment (hellofromtonya) [inline inc/Engine/HealthCheck/CacheDirSizeCheck.php:91]: Let's move the minify cache path to be a property instead that's set in the constructor. Why? If `cache_dir_size_check()` method is called more than once, the code will run a tiny bit faster as the minify path doesn't have to build again and again.

File: inc/Engine/HealthCheck/CacheDirSizeCheck.php [hunk 2]
```diff
@@ -146,14 +142,13 @@ private function get_dir_size( $dir ) {
...
 	 * @since 3.3.5
-	 * @author Remy Perona
 	 *
 	 * @param string $dir_type Type of directory.
 	 * @return void
 	 */
 	private function send_notification( $dir_type ) {
 		wp_safe_remote_post(
-			WP_ROCKET_WEB_MAIN . '/api/wp-rocket/cache-dir-check.php',
+			rocket_get_constant( 'WP_ROCKET_WEB_MAIN' ) . 'api/wp-rocket/cache-dir-check.php',
```
Comment (hellofromtonya) [inline inc/Engine/HealthCheck/CacheDirSizeCheck.php:151]: Same thing here. Let's move this to a property and set it in the constructor.

File: inc/admin/upgrader.php [hunk 3]
```diff
@@ -320,7 +319,6 @@ function rocket_new_upgrade( $wp_rocket_version, $actual_version ) {
 
 	if ( version_compare( $actual_version, '2.11', '<' ) ) {
 		rocket_clean_minify();
-		rocket_clean_cache_busting();
```
Comment (hellofromtonya) [inline inc/admin/upgrader.php:323]: What is the impact of removing this for customers running this version or older? Do we want to keep this and let it run through the deprecated functions?

File: inc/common/purge.php [hunk 4]
```diff
@@ -419,23 +419,24 @@ function rocket_post_purge_urls_for_qtranslate( $urls ) {
...
-		$type = sanitize_key( $_GET['type'] );
+		$type_raw   = sanitize_key( $_GET['type'] );
+		$type_array = explode( '-', $type_raw );
 
-		$_type     = explode( '-', $type );
-		$_type     = reset( $_type );
-		$_id       = explode( '-', $type );
-		$_id       = end( $_id );
-		$_taxonomy = isset( $_GET['taxonomy'] ) ? sanitize_title( wp_unslash( $_GET['taxonomy'] ) ) : false;
+		$type     = reset( $type_array );
```
Comment (hellofromtonya) [inline inc/common/purge.php:425]: Why is `reset()` necessary here? With the `explode()` it returns an array where the pointer is set to the first element in that array.

File: inc/common/purge.php [hunk 5]
```diff
@@ -497,50 +495,62 @@ function do_admin_post_rocket_purge_cache() { // phpcs:ignore WordPress.NamingCo
...
 				wp_nonce_ays( '' );
-				break;
+				return;
 		}
 
+		/**
+		 * Fires after the cache is cleared.
+		 *
+		 * @since  3.6
+		 * @author Grégory Viguier
```
Comment (hellofromtonya) [inline inc/common/purge.php:543]: Remove `@author`.

File: inc/deprecated/DeprecatedClassTrait.php [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+ * Trait to use in a deprecated class, or a class containing deprecated methods.
+ */
+trait DeprecatedClassTrait {
+	/**
+	 * Marks a class as deprecated and informs when it has been used.
+	 * Similar to _deprecated_constructor(), but with different strings.
+	 * The current behavior is to trigger a user error if `WP_DEBUG` is true.
+	 *
+	 * @since  3.6
+	 * @author Grégory Viguier
```
Comment (hellofromtonya) [inline inc/deprecated/DeprecatedClassTrait.php:15]: Remove all `@author` instances in this file.

File: inc/functions/admin.php [hunk 7]
```diff
@@ -76,21 +76,23 @@ function rocket_renew_box( $function, $uid = 0 ) {
...
  */
 function rocket_dismiss_box( $function ) {
-	rocket_dismiss_boxes(
-		[
-			'box'      => $function,
-			'_wpnonce' => wp_create_nonce( 'rocket_ignore_' . $function ),
-			'action'   => 'rocket_ignore',
-		]
-	);
+	global $current_user;
```
Comment (hellofromtonya) [inline inc/functions/admin.php:87]: Instead of the global, let's use `wp_get_current_user()`.

File: inc/Engine/Admin/Beacon/Beacon.php [hunk 8]
```diff
@@ -81,12 +81,12 @@ public function insert_script() {
...
-			<script>window.Beacon("session-data", ' . wp_json_encode( $this->session_data() ) . ');</script>
-			<script>window.addEventListener("hashchange", function () {
-				window.Beacon("suggest");
-			  }, false);</script>';
+		<script type="text/javascript">window.Beacon(\'init\', \'' . esc_js( $form_id ) . '\')</script>
+		<script>window.Beacon("identify", ' . wp_json_encode( $this->identify_data() ) . ');</script>
+		<script>window.Beacon("session-data", ' . wp_json_encode( $this->session_data() ) . ');</script>
+		<script>window.addEventListener("hashchange", function () {
+			window.Beacon("suggest");
+		}, false);</script>';
```
Comment (hellofromtonya) [inline inc/Engine/Admin/Beacon/Beacon.php:89]: This code is mixing concerns. I'm wondering if we want to move it to a view file. Thoughts?

File: inc/Addon/GoogleTracking/Subscriber.php [hunk 9]
```diff
@@ -13,7 +13,8 @@
  * @since 3.1
  * @author Remy Perona
```
Comment (hellofromtonya) [inline inc/Addon/GoogleTracking/Subscriber.php:14]: Remove all `@author` instances in this class.

File: inc/Addon/FacebookTracking/Subscriber.php [hunk 10]
```diff
@@ -11,7 +12,7 @@
  * @since  3.2
  * @author Grégory Viguier
```
Comment (hellofromtonya) [inline inc/Addon/FacebookTracking/Subscriber.php:13]: Remove all `@author` instances in this class.

File: inc/Engine/HealthCheck/CacheDirSizeCheck.php (hunk 1 above)
Comment (Tabrisrp) [reply inc/Engine/HealthCheck/CacheDirSizeCheck.php]: https://github.com/wp-media/wp-rocket/pull/2616/commits/9e95cf8cd5eb809236b17004a12c3b43715b4c56

File: inc/Engine/HealthCheck/CacheDirSizeCheck.php (hunk 2 above)
Comment (Tabrisrp) [reply inc/Engine/HealthCheck/CacheDirSizeCheck.php]: https://github.com/wp-media/wp-rocket/pull/2616/commits/9e95cf8cd5eb809236b17004a12c3b43715b4c56

File: inc/Addon/FacebookTracking/Subscriber.php (hunk 10 above)
Comment (Tabrisrp) [reply inc/Addon/FacebookTracking/Subscriber.php]: https://github.com/wp-media/wp-rocket/pull/2616/commits/dd041589183f9e6ea56fdbce81270869d5f90c2e

File: inc/Addon/GoogleTracking/Subscriber.php (hunk 9 above)
Comment (Tabrisrp) [reply inc/Addon/GoogleTracking/Subscriber.php]: https://github.com/wp-media/wp-rocket/pull/2616/commits/dd041589183f9e6ea56fdbce81270869d5f90c2e

File: inc/deprecated/DeprecatedClassTrait.php (hunk 6 above)
Comment (Tabrisrp) [reply inc/deprecated/DeprecatedClassTrait.php]: https://github.com/wp-media/wp-rocket/pull/2616/commits/dd041589183f9e6ea56fdbce81270869d5f90c2e

File: inc/common/purge.php (hunk 5 above)
Comment (Tabrisrp) [reply inc/common/purge.php]: https://github.com/wp-media/wp-rocket/pull/2616/commits/dd041589183f9e6ea56fdbce81270869d5f90c2e

File: inc/functions/admin.php (hunk 7 above)
Comment (Tabrisrp) [reply inc/functions/admin.php]: https://github.com/wp-media/wp-rocket/pull/2616/commits/09ba8a9c695a06ab5525478c83a5853a56e290a3

File: inc/Engine/Admin/Beacon/Beacon.php (hunk 8 above)
Comment (Tabrisrp) [reply inc/Engine/Admin/Beacon/Beacon.php]: I agree, adding a view for it makes a lot of sense. I went ahead and did that in https://github.com/wp-media/wp-rocket/pull/2616/commits/4d440f8a929f436d4a5f14009726fc98819b6ab5

File: inc/common/purge.php (hunk 4 above)
Comment (Tabrisrp) [reply inc/common/purge.php]: Removed in https://github.com/wp-media/wp-rocket/pull/2616/commits/cbd10f9a7146bc28cb754b392f427092070b9a7c

File: inc/admin/upgrader.php (hunk 3 above)
Comment (Tabrisrp) [reply inc/admin/upgrader.php:323]: Your comment makes me think we would want to run it one last time to clean the folder of the files that will no longer be used.

Comment (Tabrisrp) [reply inc/admin/upgrader.php:323]: - Clean busting directory on update: https://github.com/wp-media/wp-rocket/pull/2616/commits/352f2e22a77fd743dc78f959bd9f6567355f30af
- Undeprecate `rocket_clean_cache_busting()`: https://github.com/wp-media/wp-rocket/pull/2616/commits/c3860b6375c722dc87aa145f97df1539a5adf3f2

**Comment by hellofromtonya** [comment]
@wp-media/dev-qa This PR is ready for QA.

**Comment by Tabrisrp** [comment]
Can you share the original code used for the google tracking?

I'll see if I can replicate this, it's strange because the part handling that didn't change at all in this PR.

**Comment by arunbasillal** [comment]
@Tabrisrp This is what I used in `header.php` for GTrackign and FB Pixel - http://snippi.com/raw/z4xdafi

> it's strange because the part handling that didn't change at all in this PR.

If I remember correctly for the specific PR (#2477 ), I saw files in `\cache\busting\google-tracking` folder. But we also have the related file in `\cache\busting\1`, so I might have missed it then.

**Comment by Tabrisrp** [comment]
Thank you.

I just performed a test on my installation on the same branch, both files were generated correctly with this code, See screenshot attached.

busting

**Comment by Tabrisrp** [comment]
@hellofromtonya @crystinutzaa Requesting quick review for the changes after what Arun discovered in QA.

Commits: https://github.com/wp-media/wp-rocket/pull/2616/commits/61467bb27a664671704364ca492ee1262bab87d9 & https://github.com/wp-media/wp-rocket/pull/2616/commits/9d0e8c4c2b986d459b4e0a3c840b4fbc65a051f9

**Comment by arunbasillal** [comment]
**QA**

- Checked everything from here: https://github.com/wp-media/wp-rocket/pull/2616#pullrequestreview-408255794 all good 👍 
- Retested this: https://github.com/wp-media/wp-rocket/pull/2458#pullrequestreview-395414664 👍 
- Retested this: https://github.com/wp-media/wp-rocket/pull/2497#pullrequestreview-392900731 👍 

- Retested this: https://github.com/wp-media/wp-rocket/pull/2469#issuecomment-603821272 👍 
^ When the `busting` folder was not writeable, there was no admin notice (as per design). However if the busting folder is not writeable, Google Tracking and FB Pixel add-on doesn't work. 

Shouldn't we preserve the notice for the sake of the add-ons? (I think we should cc @GeekPress). This folder is also used for storing the minified file for SCCSS plugin (https://github.com/wp-media/wp-rocket/pull/2497#pullrequestreview-392900731)

Everything looks good from a dev standpoint. From the product pov, please let's wait for Jonathan's input. Sorry I missed that before.

**Comment by Tabrisrp** [comment]
Good catch, I re-read all the conversations in the sub-tasks, and I believe there was a mix-up when tasks were being worked on in parallel. The notice was removed on one sub-task, and on the other the discussion agreed to preserve it.

So I undeprecated the function displaying the notice to make sure we still have it.

**Pull request merged**